storage/client_range_tree_test.go

```diff
@@ -0,0 +1,379 @@
+// nodes retrieved from the db.  It recursively calls itself on both left and
+// right children if they exist.
+func treeNodesEqual(db *client.KV, expected testRangeTree, key proto.Key) error {
+	expectedNode, ok := expected.Nodes[string(key)]
+	if !ok {
+		return util.Errorf("Expected does not contain a node for %s", key)
+	}
+	actualNode := &proto.RangeTreeNode{}
+	ok, _, err := db.GetProto(engine.RangeTreeNodeKey(key), actualNode)
+	if err != nil {
+		return err
+	}
+	if !ok {
+		return util.Errorf("Could not find the range tree node for range:%s; expected:%+v", key, expectedNode)
+	}
+	if err = nodesEqual(key, expectedNode, *actualNode); err != nil {
+		return err
+	}
+	if expectedNode.LeftKey != nil {
+		if err = treeNodesEqual(db, expected, *expectedNode.LeftKey); err != nil {
+			return err
+		}
+	}
+	if expectedNode.RightKey != nil {
+		if err = treeNodesEqual(db, expected, *expectedNode.RightKey); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// treesEqual compares the expectedTree and expectedNodes to the actual range
+// tree stored in the db.
+func treesEqual(db *client.KV, expected testRangeTree) error {
+	// Compare the tree roots.
+	actualTree := &proto.RangeTree{}
+	ok, _, err := db.GetProto(engine.KeyRangeTreeRoot, actualTree)
+	if err != nil {
+		return err
+	}
+	if !ok {
+		return util.Errorf("Could not find the range tree root:%s", engine.KeyRangeTreeRoot)
+	}
+	if !reflect.DeepEqual(&expected.Tree, actualTree) {
+		return util.Errorf("Range tree root is not as expected - expected:%+v - actual:%+v", expected.Tree, actualTree)
+	}
+
+	return treeNodesEqual(db, expected, expected.Tree.RootKey)
+}
+
+// splitRange splits whichever range contains the split key on the split key.
+func splitRange(db *client.KV, key proto.Key, splitKey proto.Key) error {
+	req := &proto.AdminSplitRequest{
+		RequestHeader: proto.RequestHeader{
+			Key: key,
+		},
+		SplitKey: splitKey,
+	}
+	resp := &proto.AdminSplitResponse{}
+	if err := db.Call(proto.AdminSplit, req, resp); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// TestSetupRangeTree ensures that SetupRangeTree correctly setups up the range
+// tree and first node.  SetupRangeTree is called via store.BootstrapRange.
+func TestSetupRangeTree(t *testing.T) {
+	defer leaktest.AfterTest(t)
+	store := createTestStore(t)
+	defer store.Stop()
+
+	// Check to make sure the range tree is stored correctly.
+	tree := proto.RangeTree{
+		RootKey: engine.KeyMin,
+	}
+	nodes := map[string]proto.RangeTreeNode{
+		string(engine.KeyMin): proto.RangeTreeNode{
+			Key:       engine.KeyMin,
+			ParentKey: engine.KeyMin,
+			Black:     true,
+		},
+	}
+	expectedTree := testRangeTree{
+		Tree:  tree,
+		Nodes: nodes,
+	}
+	if err := treesEqual(store.DB(), expectedTree); err != nil {
+		t.Fatal(err)
+	}
+}
+
+// TestInsert tests inserting a collection of 5 nodes, forcing left rotations
+// and flips.
+func TestInsert(t *testing.T) {
```

Comment:
I agree entirely.  I'm going to keep this test in and actually create a mirror test that does left splitting each time (parent node will always be MinKey).

With the addition of the cache, it means I should be able to test the tree in isolation without needing to roll up a full cluster, so with that, i can mimic a lot of the tests from the biogo's library.

But I do like the idea of randomly performing inserts in both trees and comparing them.  I'll add that test in as well.